taichi/ir/frontend_ir.h

```diff
@@ -521,6 +521,24 @@ class FieldExpression : public Expression {
   TI_DEFINE_ACCEPT_FOR_EXPRESSION
 };
 
+class MatrixFieldExpression : public Expression {
+ public:
+  std::vector<Expr> fields;
+  std::vector<int> element_shape;
```

Comment:
I was thinking we can infer `element_shape` from `ret_type`, whereas `ret_type` for MatrixFieldExpression should have already been set upon construction?

Reply:
This part is a bit tricky - for some use cases (like quant), the fields may have different dtypes, so the result type cannot be simply represented as `TensorType`. I plan to scalarize such tricky cases during `lower_ast`. After `lower_ast` things will work as we discussed.